clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -48,11 +48,21 @@ scheduler_tmpl_t<RecordType, ReaderType>::init(
     std::vector<input_workload_t> &workload_inputs, int output_count,
     scheduler_options_t options)
 {
-    // TODO i#6831: Split up scheduler_impl_tmpl_t by mapping_t mode and create the
-    // mode-appropriate subclass here.
-    impl_ = std::unique_ptr<scheduler_impl_tmpl_t<RecordType, ReaderType>,
-                            scheduler_impl_deleter_t>(
-        new scheduler_impl_tmpl_t<RecordType, ReaderType>);
+    if (options.mapping == sched_type_t::MAP_TO_ANY_OUTPUT) {
+        impl_ = std::unique_ptr<scheduler_impl_tmpl_t<RecordType, ReaderType>,
+                                scheduler_impl_deleter_t>(
+            new scheduler_dynamic_tmpl_t<RecordType, ReaderType>);
+    } else if (options.mapping == sched_type_t::MAP_AS_PREVIOUSLY ||
+               (options.mapping == sched_type_t::MAP_TO_RECORDED_OUTPUT &&
+                options.replay_as_traced_istream != nullptr)) {
+        impl_ = std::unique_ptr<scheduler_impl_tmpl_t<RecordType, ReaderType>,
+                                scheduler_impl_deleter_t>(
+            new scheduler_replay_tmpl_t<RecordType, ReaderType>);
+    } else {
+        impl_ = std::unique_ptr<scheduler_impl_tmpl_t<RecordType, ReaderType>,
+                                scheduler_impl_deleter_t>(
+            new scheduler_fixed_tmpl_t<RecordType, ReaderType>);
```

Comment:
Is this intended to be only for MAP_TO_CONSISTENT_OUTPUT? Or also a catch-all for things like MAP_TO_CONSISTENT_OUTPUT but without replay_as_traced_istream?

Maybe we should check for bad configs (e..g, MAP_TO_CONSISTENT_OUTPUT without replay_as_traced_istream) here and fail early instead of waiting for the first call to pick_next_input_try? Atleast for config errors that can be determined here itself.

Reply:
The latter: anything that's not dynamic and is not replay, which generally is serial or parallel analyzer modes.

Bad configs are checked in scheduler_impl_tmpl_t::init() and its helpers like set_initial_schedule() (which will probably be split by mode in the future): so they are checked right up front in the same user init() call.

Reply:
Added a comment here similar to the class-level one for _fixed_:

        // Non-dynamic and non-replay fixed modes such as analyzer serial and
        // parallel modes.